cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/table/TableSpecification.java

```diff
@@ -0,0 +1,201 @@
+  }
+
+  public TableId getTableId() {
+    return tableId;
+  }
+
+  public List<FieldType> getFields() {
+    return fields;
+  }
+
+  public List<String> getPrimaryKeys() {
+    return primaryKeys;
+  }
+
+  public List<String> getIndexes() {
+    return indexes;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    TableSpecification that = (TableSpecification) o;
+    return Objects.equals(tableId, that.tableId) &&
+      Objects.equals(fields, that.fields) &&
+      Objects.equals(primaryKeys, that.primaryKeys) &&
+      Objects.equals(indexes, that.indexes);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(tableId, fields, primaryKeys, indexes);
+  }
+
+  @Override
+  public String toString() {
+    return "TableSpecification{" +
+      "tableId='" + tableId + '\'' +
+      ", fields=" + fields +
+      ", primaryKeys=" + primaryKeys +
+      ", indexes=" + indexes +
+      '}';
+  }
+
+  /**
+   * Builder used to create {@link TableSpecification}
+   */
+  public static final class Builder {
+    private TableId tableId;
+    private FieldType[] fieldTypes;
+    private String[] primaryKeys;
+    private String[] indexes;
+
+    public Builder withId(TableId id) {
+      this.tableId = id;
+      return this;
+    }
+
+    public Builder withFields(FieldType ...fieldTypes) {
+      this.fieldTypes = fieldTypes;
+      return this;
+    }
+
+    public Builder withPrimaryKeys(String ...primaryKeys) {
+      this.primaryKeys = primaryKeys;
+      return this;
+    }
+
+    public Builder withIndexes(String ...indexes) {
+      this.indexes = indexes;
+      return this;
+    }
+
+    public TableSpecification build() {
+      validate();
+      return new TableSpecification(tableId, Arrays.asList(fieldTypes), Arrays.asList(primaryKeys),
+                                    Arrays.asList(indexes));
+    }
+
+    private void validate() {
+      if (tableId == null) {
+        throw new IllegalArgumentException("TableId cannot be empty");
+      }
+
+      // Validate the table name is made up of valid characters
+      if (!IDENTIFIER_NAME_PATTERN.matcher(tableId.getName()).matches()) {
+        throw new IllegalArgumentException(
+          String.format(
+            "Invalid table name %s. Only alphanumeric and _ characters allowed, and should begin with an alphabet",
+            tableId.getName()));
+      }
+
+      if (fieldTypes == null || fieldTypes.length == 0) {
+        throw new IllegalArgumentException("No fields specified for the table " + tableId);
+      }
+
+      if (primaryKeys == null || primaryKeys.length == 0) {
+        throw new IllegalArgumentException("No primary keys specified for the table " + tableId);
+      }
+
+      // A table need not have indexes
+      if (indexes == null) {
+        indexes = new String[0];
```

Comment:
It's better not to modify field in a validate method. I think it's better to have the field initialized as empty string when it was declared and have the `withXXX` method to fail if a `null` array is being passed in.